### PR TITLE
Fixes #27934 - Initial Content units migration task

### PR DESCRIPTION
**Pull request by jlsherrill**
This provides the ability to migrate arbitrary
content units (file_units, docker tags, etc..) from
pulp2 to pulp3 and import the pulp3 to our database

### CODE REVIEW

**Comment by jlsherrill** [comment]
Ready for review, but a few TODOs:

* [ ] - test with a large file repo
* [ ] - package the pulp2to3 gem in foremanpackaging

**Comment by jlsherrill** [comment]
How to test:
* rake db:migrate
* sync a bunch of file repos with pulp2 (might need to use the new setting in https://github.com/Katello/katello/commit/340868d4de24929897611d536b4a86c5b79fe4ef)
* run 'rake katello:pulp3_migration
* Verify that:  `Katello::FileUnit.where(:migrated_pulp3_href => nil)` returns nothing

**Comment by ianballou** [comment]
I got:
```
pulp: django.request:WARNING: Not Found: /pulp/api/v3/migration-plans/
127.0.0.1 - admin [17/Oct/2019:20:28:17 +0000] "POST /pulp/api/v3/migration-plans/ HTTP/1.1" 404 77 "-" "OpenAPI-Generator/0.0.1a1.dev01571338032/ruby"
```
Perhaps my pulp is out of date?

**Comment by jlsherrill** [comment]
@ianballou do you have the pulp2to3 migration plugin installed?  My latest forklift pr should do that for you

**Comment by jlsherrill** [comment]
rebased

**Comment by ianballou** [comment]
I got:
```
pulp: django.request:WARNING: Not Found: /pulp/api/v3/migration-plans/
127.0.0.1 - admin [17/Oct/2019:20:28:17 +0000] "POST /pulp/api/v3/migration-plans/ HTTP/1.1" 404 77 "-" "OpenAPI-Generator/0.0.1a1.dev01571338032/ruby"
```
Perhaps my pulp is out of date?

> @ianballou do you have the pulp2to3 migration plugin installed? My latest forklift pr should do that for you

My bad, this is the issue.  Testing again.

**Comment by ianballou** [comment]
I can hit the API now, but the Pulp server is giving me a 500.  Here's the [Pulp 3 log](https://paste.fedoraproject.org/paste/SX74jcigCiVeClIc78nOKQ) and here's the matching [Katello log](https://paste.fedoraproject.org/paste/nUKlfVICh-pJ7xs6CzbVVA).

**Comment by jjeffers** [comment]
@jlsherrill I ran this migration with no pulp2 content at all and the task didn't seem to be doing anything, nor did it exit. Just FYI.

File: app/services/katello/pulp3/migration.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        @repository_type_labels.each do |repository_type_label|
+          Katello::RepositoryTypeManager.repository_types[repository_type_label].content_types_to_index.each do |content_type|
+            import_content_type(content_type)
+          end
+        end
+      end
+
+      def create_migration
+        migration_plan = Katello::Pulp3::MigrationPlan.new(@repository_type_labels)
+        migration_plan_api.create(plan: migration_plan.generate.to_json)._href
```
Comment (ianballou) [inline app/services/katello/pulp3/migration.rb]: `%s/_href/pulp_href`

**Comment by jlsherrill** [comment]
@ianballou good question.  Pulp should only execute one at a time, so its probably not a big deal if multiple run at the same time in katello.  I'm not actually sure how we would lock it, as locks are handled via Active record objects, and we don't have a singular object to lock on

File: app/lib/actions/pulp3/content_migration.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module Actions
+  module Pulp3
+    class ContentMigration < Pulp3::AbstractAsyncTask
+      def plan(repository_type_labels, smart_proxy = SmartProxy.pulp_master)
+        sequence do
+          plan_self(repository_type_labels: repository_type_labels, :smart_proxy_id => smart_proxy.id)
```
[the remark this answers is not in the document]
Comment (jlsherrill) [reply app/lib/actions/pulp3/content_migration.rb]: Good catch!

**Pull request merged**